statie/source/_posts/2019/2019-05-21-review-of-the-refactoring.md

```diff
@@ -0,0 +1,80 @@
+---
+id: 82
+title: "Book Review: Refactoring"
+perex: |
+    Review of the Refactoring: Improving the Design of Existing Code by [Martin Fowler](https://martinfowler.com/), 2nd edition (2018).
+author: 29
+tweet: "New Post on #pehapkari Blog: Review of the Refactoring book #refactoring #martinfowler"
+---
+
+![Refactoring book](https://martinfowler.com/books/refact2.jpg)
+
+The book starts right away with an example of low-quality code and our goal is to add a feature.
+Fowler attracts our attention as we immediately realize the need for refactoring.
+Then we apply one refactoring after another, improve the code structure and after a couple of pages, we are able to add the feature without any additional effort.
+The code is simply prepared for the feature to be added.
+
+The author explains the importance of refactoring in a natural way - we realize we need to refactor, there is no other option.
+Also often emphasizes the importance of automated tests because we cannot refactor without a proper test suite.
+The first step of refactoring is to create tests and the book explains how to start with tests (but there are better sources for starting with tests like [TDD by Kent Beck](https://www.amazon.com/Test-Driven-Development-Kent-Beck/dp/0321146530?tag=uuid10-20)).
+
+Honestly, the important part of the book is the first 100 pages (of 400) and I recommend for every developer to read the first part of the book, the rest of the book is optional.
+
+## Catalogue of Refactorings
+
+The rest 300 pages is a catalog of refactoring and you can find it [online](https://refactoring.com/catalog/).
+Each refactoring includes motivation - when, mechanics - how - in step-by-step form, and examples.
+I like the step-by-step instructions when tests pass after each instruction.
+It is awkward to apply these steps for simple refactorings, and powerful to apply for complicated refactorings or refactorings that involves a lot of occurrences.
+
+All refactorings are explained really simply, even the difficult parts.
+That's not because oversimplification, but it's because the author is a professional teacher, the explanation is just great.
+Most of the refactorings are simple (sometimes too much like [Remove Dead Code](https://refactoring.com/catalog/removeDeadCode.html)) but still practical.
+
+## Personal Impression
+
+Almost all refactorings include inverse refactoring.
+When there is [Extract Variable](https://refactoring.com/catalog/extractVariable.html), you can find also [Inline Variable](https://refactoring.com/catalog/inlineVariable.html).
+That was a surprise for me as I thought the first option is always better (and I learned that's not true).
```

Comment:
That's intersting Could you maybe explain why? + 2 examples when better extract and better inline?

It would give reader more value he or she could use, than just bar statement.